### PR TITLE
Support multiple crypto curves

### PR DESCRIPTION
**Pull request by eddyashton**
Previously we only built a single implementation in `keypair.h`, so all cryptographic signing/verification (whether for the ledger or for TLS or app-specific) was forced to use the same curve. This PR removes that restriction - support for all (currently 3) curves is always built, and different parts of the code can generate keys or verify any of these curves.

Additionally, we support 2 different implementations of the `secp256k1` curve, and we can use both implementations in the same program. To achieve this, I've changed the `secp256k1_bitcoin` functions to produce and verify non-recoverable signatures. This way it is compatible with signatures produced by other `secp256k1` implementations (ie - the mbedtls version). I'll re-add the recoverable signatures where we need it - that may be app-local.

* The bulk of the changes are in `keypair.h`. I've tried to rationalise and simplify it, so the various `sign` and `verify` generally forward arguments to a single place that makes the `mbedtls_*` calls, rather than duplicating an opaque function sequence.

* Rather than constructing a `KeyPair`/`PublicKey`/`Verifier` directly we should (generally) now use the `tls::make_*` factory functions. The `sign`/`verify` methods on the resulting object are virtual, allowing different implementations and curves to be plugged in transparently. Many of the changes are simply doing this swap throughout the codebase.

* `CURVE_CHOICE` is replaced by `LEDGER_CURVE_CHOICE`. This sets the curve which should be used for the network and node identities. For simplicity this is also the default curve, used for anything which isn't more explicit.

* `tls_test` is extended to run its tests against all supported curves, and also to confirm that we can verify `secp256k1_mbedtls` signatures with `secp256k1_bitcoin` code and vice versa.

* `tls_bench` compares all supported curves, showing the `LEDGER_CURVE_CHOICE` as the baseline.

* The calls to `sign_hash` and `verify_hash` have changed slightly. I think these are still correct, but since they're at the heart of the merkle tree history they should be double-checked. I think some uses were previously double-hashing data unnecessarily. Generally we should be passing the actual contents, and `sign` will hash and sign them correctly. If we think we have pre-hashed data from elsewhere, we need to ensure it is of the correct _size_ for the target's signing algorithm.

### CODE REVIEW

File: src/tls/keypair.h [hunk 1]
```diff
@@ -17,67 +17,159 @@
...
+        constexpr auto hash_size = 384 / 8;
+        if (o_hash.size() < hash_size)
+          o_hash.resize(hash_size);
 
-  static constexpr size_t REC_ID_IDX = 64;
+        return mbedtls_sha512_ret(data_ptr, data_size, o_hash.data(), true);
+      }
+      case MBEDTLS_ECP_DP_CURVE25519:
+      {
+        constexpr auto hash_size = 512 / 8;
```
Comment (jumaffre) [inline src/tls/keypair.h]: Not really part of this refactoring but I think we could retrieve this from `mbedtls` directly, e.g. https://github.com/ARMmbed/mbedtls/blob/8266acacc8d6e1c65fba9a048f56339d0827b2fe/library/md_wrap.c#L572 ?

File: src/tls/keypair.h [hunk 2]
```diff
@@ -580,145 +616,191 @@ namespace tls
...
+        k1_ctx, sig, sig_size, hash, hash_size, &c4_pub);
+    }
+  };
+
+  using PublicKeyPtr = std::shared_ptr<PublicKey>;
+
+  /**
+   * Construct PublicKey from a raw public key in PEM format
+   *
+   * @param public_pem Sequence of bytes containing the key in PEM format
```
Comment (jumaffre) [inline src/tls/keypair.h]: Minor: `@param use_bitcoin_impl` missing

File: src/tls/keypair.h [hunk 3]
```diff
@@ -17,67 +17,159 @@
...
+{
+  enum class CurveImpl
+  {
+    secp384r1 = 1,
+    curve25519 = 2,
+    secp256k1_mbedtls = 3,
+    secp256k1_bitcoin = 4,
+
+#if LEDGER_CURVE_CHOICE_SECP384R1
+    ledger_curve_choice = secp384r1,
```
Comment (jumaffre) [inline src/tls/keypair.h]: There's something that I'm probably missing but I'm not sure the curve choice should mention "ledger" as it is used in a bunch of places that are not directly related to the ledger itself (e.g. node and network identity).

File: src/tls/test/main.cpp [hunk 4]
```diff
@@ -22,67 +22,205 @@ static const string contents_ =
...
+  buf[buf.size() - 2]++;
 }
 
+static constexpr tls::CurveImpl supported_curves[] = {
+  tls::CurveImpl::secp384r1,
+  tls::CurveImpl::curve25519,
+  tls::CurveImpl::secp256k1_mbedtls,
+  tls::CurveImpl::secp256k1_bitcoin};
+
+static constexpr char const* labels[] = {
```
Comment (jumaffre) [inline src/tls/test/main.cpp]: I guess this could be moved to `keypair.h`?

File: src/tls/test/main.cpp [hunk 5]
```diff
@@ -22,67 +22,205 @@ static const string contents_ =
...
+    vector<uint8_t> signature = kp->sign(contents);
+
+    auto kp2 = tls::make_key_pair(curve);
+    vector<uint8_t> public_key = kp2->public_key();
+    auto pubk = tls::make_public_key(public_key);
+    CHECK_FALSE(pubk->verify(contents, signature));
+  }
+}
+
+TEST_CASE("Sign, fail to verify with wrong key on wrong curve")
```
Comment (jumaffre) [inline src/tls/test/main.cpp]: Should we add a test for "wrong key on right curve"?

File: cmake/common.cmake [hunk 6]
```diff
@@ -44,17 +44,17 @@ function(enable_coverage name)
...
-set(CURVE_CHOICE "secp384r1" CACHE STRING "One of secp384r1, curve25519, secp256k1_mbedtls, secp256k1_bitcoin")
-if (${CURVE_CHOICE} STREQUAL "secp384r1")
-  add_definitions(-DCURVE_CHOICE_SECP384R1)
-elseif (${CURVE_CHOICE} STREQUAL "curve25519")
-  add_definitions(-DCURVE_CHOICE_CURVE25519)
-elseif (${CURVE_CHOICE} STREQUAL "secp256k1_mbedtls")
-  add_definitions(-DCURVE_CHOICE_SECP256K1_MBEDTLS)
-elseif (${CURVE_CHOICE} STREQUAL "secp256k1_bitcoin")
-  add_definitions(-DCURVE_CHOICE_SECP256K1_BITCOIN)
+set(LEDGER_CURVE_CHOICE "secp384r1" CACHE STRING "One of secp384r1, curve25519, secp256k1_mbedtls, secp256k1_bitcoin")
```
Comment (achamayou) [inline cmake/common.cmake]: Why ledger curve? The ledger is encrypted with symmetric encryption, this is only for identities.

File: sphinx/source/getting_started.rst [hunk 7]
```diff
@@ -163,7 +163,7 @@ The full list of build switches can be obtained by running:
 * **BUILD_TESTS**: Boolean. Build all tests for CCF. Default to ON.
 * **BUILD_SMALLBANK**: Boolean. Build SmallBank performance benchmark. Default to OFF.
 * **CLIENT_MBEDTLS_PREFIX**: Path. Prefix to mbedtls install to be used by test clients. Default to ``/usr/local``.
-* **CURVE_CHOICE**: String, one of ``secp384r1``, ``curve25519``, ``secp256k1_mbedtls``, ``secp256k1_bitcoin``. Elliptic curve to use for CCF asymmetric keys. Default to ``secp384r1``.
+* **LEDGER_CURVE_CHOICE**: String, one of ``secp384r1``, ``curve25519``, ``secp256k1_mbedtls``, ``secp256k1_bitcoin``. Elliptic curve to use for CCF network and node identity. Defaults to ``secp384r1``.
```
Comment (achamayou) [inline sphinx/source/getting_started.rst]: Service identity curve perhaps?

File: src/node/rpc/test/membervoting_test.cpp [hunk 8]
```diff
@@ -290,9 +290,9 @@ TEST_CASE("Add new members until there are 7, then reject")
     const auto proposal_id = i;
     new_member.id = initial_members + i++;
     // new member certificate
-    tls::Verifier v(
-      new_member.kp.self_sign("CN=new member" + to_string(new_member.id)));
-    const auto _cert = v.raw();
+    auto v = tls::make_verifier(
+      new_member.kp->self_sign("CN=new member" + to_string(new_member.id)));
```
Comment (achamayou) [inline src/node/rpc/test/membervoting_test.cpp]: fmt::format("CN=new member{}", new_member.id)

File: src/tls/keypair.h [hunk 9]
```diff
@@ -164,54 +240,23 @@ namespace tls
...
-    {
-      key = std::move(other.key);
-      other.key = nullptr;
-#if CURVE_CHOICE_SECP256K1_BITCOIN
-      ctx = std::move(other.ctx);
-      other.ctx = nullptr;
-#endif
+      if (get_ec_from_context(*key) != ec)
+      {
+        throw std::logic_error("Created key and received unexpected type");
```
Comment (achamayou) [inline src/tls/keypair.h]: Might be good to print the mismatching value as well.

File: src/tls/keypair.h [hunk 10]
```diff
@@ -501,43 +470,128 @@ namespace tls
...
+    bool use_bitcoin_impl = prefer_bitcoin_secp256k1)
+  {
+    std::unique_ptr<mbedtls_pk_context> key =
+      std::make_unique<mbedtls_pk_context>();
+    mbedtls_pk_init(key.get());
+
+    Pem pemPk(pkey);
+    if (mbedtls_pk_parse_key(key.get(), pemPk.p, pemPk.n, pw.p, pw.n) != 0)
+    {
+      throw std::logic_error("Could not parse key");
```
Comment (achamayou) [inline src/tls/keypair.h]: Including the rc would be good.

File: src/tls/keypair.h [hunk 11]
```diff
@@ -501,43 +470,128 @@ namespace tls
...
+   */
+  template <typename... Ts>
+  inline KeyPairPtr make_key_pair(
+    CBuffer pkey,
+    CBuffer pw = nullb,
+    bool use_bitcoin_impl = prefer_bitcoin_secp256k1)
+  {
+    std::unique_ptr<mbedtls_pk_context> key =
+      std::make_unique<mbedtls_pk_context>();
+    mbedtls_pk_init(key.get());
```
Comment (achamayou) [inline src/tls/keypair.h]: can this fail?

File: src/tls/keypair.h [hunk 12]
```diff
@@ -501,43 +470,128 @@ namespace tls
...
+    }
+  };
+
+  using KeyPairPtr = std::shared_ptr<KeyPair>;
+
+  /**
+   * Create a new public / private key pair on specified curve and
+   * implementation
+   */
+  template <typename... Ts>
```
Comment (achamayou) [inline src/tls/keypair.h]: Why the parameter pack?

File: src/tls/keypair.h [hunk 13]
```diff
@@ -501,43 +470,128 @@ namespace tls
...
+    else
+    {
+      return KeyPairPtr(new KeyPair(ec));
+    }
+  }
+
+  /**
+   * Create a public / private from existing raw private key data
+   */
+  template <typename... Ts>
```
Comment (achamayou) [inline src/tls/keypair.h]: Why the parameter pack?

File: src/tls/keypair.h [hunk 14]
```diff
@@ -751,13 +833,87 @@ namespace tls
...
+      secp256k1_context_create(SECP256K1_CONTEXT_VERIFY);
+    secp256k1_pubkey c4_pub;
+
+  public:
+    template <typename... Ts>
+    Verifier_k1Bitcoin(Ts... ts) : Verifier(std::forward<Ts>(ts)...)
+    {
+      auto k = mbedtls_pk_ec(cert.pk);
+      size_t pub_len;
+      uint8_t pub_buf[100];
```
Comment (achamayou) [inline src/tls/keypair.h]: Where does 100 come from?

File: cmake/common.cmake (hunk 6 above)
Comment (eddyashton) [reply cmake/common.cmake]: I don't know why I ended up with `LEDGER` here, `SERVICE_IDENTITY` is much better.

File: src/tls/test/main.cpp (hunk 5 above)
Comment (eddyashton) [reply src/tls/test/main.cpp]: That's the test case above this one

`TEST_CASE("Sign, fail to verify with wrong key on correct curve")`

File: src/tls/test/main.cpp (hunk 4 above)
Comment (eddyashton) [reply src/tls/test/main.cpp]: To get labels like this we need to do the ugly `static_cast` from the `enum class` so I'd prefer to keep it local to the tests, where its only used for debug information. If we need labels elsewhere, we can add a more explicit solution to keypair.h.

File: src/tls/keypair.h (hunk 1 above)
Comment (eddyashton) [reply src/tls/keypair.h]: Good spot. Turns out remove this entire switch and let mbedtls find the sizes and algorithm for us.

File: src/tls/keypair.h (hunk 9 above)
Comment (eddyashton) [reply src/tls/keypair.h]: I've now added the return codes/unexpected values to all logic errors thrown from `keypair.h`.

File: src/tls/keypair.h (hunk 12 above)
Comment (eddyashton) [reply src/tls/keypair.h]: This was previously forwarding additional arguments. It's not now - removed!

File: src/tls/keypair.h (hunk 11 above)
Comment (eddyashton) [reply src/tls/keypair.h]: [Documentation ](https://tls.mbed.org/api/pk_8h.html#a999d1160bb30c03d0c4382c3a9b0aa89) doesn't suggest so. Return type is void.

File: src/tls/keypair.h (hunk 14 above)
Comment (eddyashton) [reply src/tls/keypair.h]: Line 650 in the previous version. There's nothing obvious in the mbedtls documentation giving a suggested size here. I'll give the magic constant a name though.

Comment (eddyashton) [reply src/tls/keypair.h]: And since this is the same initialisation needed for `PublicKey_k1Bitcoin`, I've extracted it to `parse_secp256k_bc`
